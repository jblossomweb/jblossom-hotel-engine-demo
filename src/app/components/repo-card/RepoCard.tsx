import React from 'react';
import { Card, Button } from 'antd';
import { Link } from 'react-router-dom';
import {
  GithubOutlined,
  StarOutlined,
  CodeOutlined,
  ForkOutlined,
  RightCircleOutlined,
} from '@ant-design/icons';
import { Repository, User } from '../../types';
import { FlexRowLeft } from '../../shared.style';

export interface Props {
  owner: User['login'];
  name: Repository['name'];
  description: Repository['description'];
  stargazers_count: Repository['stargazers_count'];
  forks_count: Repository['forks_count'];
  language: Repository['language'];
}

const RepoCard: React.FC<Props> = ({
  name,
  description,
  stargazers_count,
  forks_count,
  language,
  owner,
}) => (
  <Card>
    <h3>
      <GithubOutlined />
      &nbsp;
      <Link to={`/repos/${owner}/${name}`}>
        {owner}/{name}
      </Link>
    </h3>
    <p>{description}</p>
    <FlexRowLeft>
      {language?.length ? (
        <p>
          <CodeOutlined />
          &nbsp;
          {language}
        </p>
      ) : null}
      <p>
        <StarOutlined />
        &nbsp;
        {stargazers_count.toLocaleString()}
      </p>
      <p>
        <ForkOutlined />
        &nbsp;
        {forks_count.toLocaleString()}
      </p>
    </FlexRowLeft>
    <FlexRowLeft style={{ marginBottom: 16 }}>
      <Button
        type="primary"
        href={`/repos/${owner}/${name}`}
        icon={<RightCircleOutlined />}
      >
        Details
      </Button>
    </FlexRowLeft>
  </Card>
);

export default RepoCard;
