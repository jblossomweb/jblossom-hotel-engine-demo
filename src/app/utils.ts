export const requestTimeout = (ms: number) =>
  new Promise((_resolve, reject) => {
    const wait = setTimeout(() => {
      clearTimeout(wait);
      reject(new Error('Request Timeout'));
    }, ms);
  });

export const trimWords = (words: string) =>
  words
    .split(' ')
    .filter((word) => word.length)
    .map((word) => word.trim())
    .join(' ');
