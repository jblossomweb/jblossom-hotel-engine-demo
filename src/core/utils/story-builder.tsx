/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import * as knobs from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import 'antd/dist/antd.css';
import Theme from '../../app/theme.style';

const { withKnobs } = knobs;

export interface KnobsInterface {
  text: (name: string, val: string) => string;
  number: (name: string, val: number) => number;
  date: (name: string, val: number) => number;
  color: (name: string, val: string) => string;
  boolean: (name: string, val: boolean) => boolean;
  object: (name: string, val: any) => any;
  select: (name: string, options: any, val: any) => any;
}

export const mockKnobs: KnobsInterface = {
  text: (name: string, val: string) => val,
  number: (name: string, val: number) => val,
  date: (name: string, val: number) => val,
  color: (name: string, val: string) => val,
  boolean: (name: string, val: boolean) => val,
  object: (name: string, val: object) => val,
  select: (name: string, options: any, val: any) => val,
};

export type Story = (...args: any[]) => React.ReactElement;

export interface Stories {
  [key: string]: Story;
}

export const storyWrapper = (stories: Stories) => {
  const wrappedStories: Stories = {};
  Object.keys(stories).forEach((key) => {
    wrappedStories[key] = () => (
      <>
        <Theme />
        <MemoryRouter>{stories[key](knobs)}</MemoryRouter>
      </>
    );
  });
  return wrappedStories;
};

export const storyBuilder = (stories: Stories, storyPath: string) => {
  const wrappedStories: Stories = storyWrapper(stories);
  const builtStories = storiesOf(storyPath, module);
  builtStories.addDecorator(withKnobs as any);
  Object.keys(wrappedStories).forEach((key) => {
    builtStories.add(key, wrappedStories[key]);
  });
};
