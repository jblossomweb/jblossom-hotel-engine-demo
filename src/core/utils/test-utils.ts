export const randomNumber = (min: number, max: number) =>
  Math.random() * (max - min) + min;

export const randomInteger = (min: number, max: number) =>
  Math.floor(randomNumber(min, max));

export const randomString = (length: number) =>
  Array(length)
    .fill(0)
    .map(() => Math.random().toString(36).charAt(2))
    .join('');
