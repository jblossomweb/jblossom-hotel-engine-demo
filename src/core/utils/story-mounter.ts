/* eslint-disable @typescript-eslint/no-explicit-any */
import './enzyme-adapter';
import { shallow, mount, ReactWrapper } from 'enzyme';
import { Stories, mockKnobs } from './story-builder';

export interface MountedStory {
  wrapper: ReactWrapper;
  component: ReactWrapper;
  props: any;
  elements: { [key: string]: ReactWrapper<any> };
}

const storyMounter = (
  stories: Stories,
  mainComponent: React.Component<any> | React.FC<any>,
  getElements?: (component: ReactWrapper) => MountedStory['elements'],
  mountShallow?: boolean
) => {
  const mountedStories: { [key: string]: MountedStory } = {};
  // tslint:disable-next-line: ban-types
  const mounter: Function = mountShallow ? shallow : mount;
  Object.keys(stories).forEach((key: string) => {
    const story = stories[key];
    const wrapper: ReactWrapper = mounter(story(mockKnobs));
    const component: ReactWrapper = wrapper.find(mainComponent);
    const props = component.props();
    const elements = getElements ? getElements(component) : {};
    mountedStories[key] = {
      wrapper,
      component,
      props,
      elements,
    };
  });
  return mountedStories;
};

export default storyMounter;
