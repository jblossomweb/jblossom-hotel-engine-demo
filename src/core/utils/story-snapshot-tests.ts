import renderer from 'react-test-renderer';
import { Story, Stories, mockKnobs } from './story-builder';

export const renderTree = (story: Story, classesWithRefs?: string[]) =>
  renderer
    .create(story(mockKnobs), {
      createNodeMock: (element) => {
        // optional workaround for known bug with react-test-renderer and refs
        if (classesWithRefs?.includes(element.props.className)) {
          return {
            scrollWidth: 0,
          };
        }
        return null;
      },
    })
    .toJSON();

const snapshotTests = (stories: Stories, classesWithRefs?: string[]) => {
  Object.keys(stories).forEach((key: string) => {
    const story = stories[key];
    describe(`${key} story`, () => {
      it('matches snapshot', () => {
        const tree = renderTree(story, classesWithRefs);
        expect(tree).toMatchSnapshot();
      });
    });
  });
};

export default snapshotTests;
