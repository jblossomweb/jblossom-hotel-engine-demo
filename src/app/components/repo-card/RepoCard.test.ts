import snapshotTests from '../../../core/utils/story-snapshot-tests';
import { storyWrapper } from '../../../core/utils/story-builder';
import { stories } from './RepoCard.stories';

const wrappedStories = storyWrapper(stories);

describe('components/repo-card', () => {
  snapshotTests(wrappedStories);
});
