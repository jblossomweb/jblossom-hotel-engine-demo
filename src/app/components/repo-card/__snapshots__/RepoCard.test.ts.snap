// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/repo-card empty language story matches snapshot 1`] = `
<div
  className="ant-card ant-card-bordered"
>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <h3>
      <span
        aria-label="github"
        className="anticon anticon-github"
        role="img"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="github"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0138.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z"
          />
        </svg>
      </span>
       
      <a
        href="/repos/freeCodeCamp/freeCodeCamp"
        onClick={[Function]}
      >
        freeCodeCamp
        /
        freeCodeCamp
      </a>
    </h3>
    <p>
      freeCodeCamp.org's open source codebase and curriculum. Learn to code at home.
    </p>
    <div
      className="sc-AxirZ qsqry"
    >
      <p>
        <span
          aria-label="star"
          className="anticon anticon-star"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="star"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
            />
          </svg>
        </span>
         
        314,981
      </p>
      <p>
        <span
          aria-label="fork"
          className="anticon anticon-fork"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="fork"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M752 100c-61.8 0-112 50.2-112 112 0 47.7 29.9 88.5 72 104.6v27.6L512 601.4 312 344.2v-27.6c42.1-16.1 72-56.9 72-104.6 0-61.8-50.2-112-112-112s-112 50.2-112 112c0 50.6 33.8 93.5 80 107.3v34.4c0 9.7 3.3 19.3 9.3 27L476 672.3v33.6c-44.2 15-76 56.9-76 106.1 0 61.8 50.2 112 112 112s112-50.2 112-112c0-49.2-31.8-91-76-106.1v-33.6l226.7-291.6c6-7.7 9.3-17.3 9.3-27v-34.4c46.2-13.8 80-56.7 80-107.3 0-61.8-50.2-112-112-112zM224 212a48.01 48.01 0 0196 0 48.01 48.01 0 01-96 0zm336 600a48.01 48.01 0 01-96 0 48.01 48.01 0 0196 0zm192-552a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
            />
          </svg>
        </span>
         
        24,727
      </p>
    </div>
    <div
      className="sc-AxirZ qsqry"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <a
        className="ant-btn ant-btn-primary"
        href="/repos/freeCodeCamp/freeCodeCamp"
        onClick={[Function]}
      >
        <span
          aria-label="right-circle"
          className="anticon anticon-right-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="right-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M666.7 505.5l-246-178A8 8 0 00408 334v46.9c0 10.2 4.9 19.9 13.2 25.9L566.6 512 421.2 617.2c-8.3 6-13.2 15.6-13.2 25.9V690c0 6.5 7.4 10.3 12.7 6.5l246-178c4.4-3.2 4.4-9.8 0-13z"
            />
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
          </svg>
        </span>
        <span>
          Details
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`components/repo-card sample story matches snapshot 1`] = `
<div
  className="ant-card ant-card-bordered"
>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <h3>
      <span
        aria-label="github"
        className="anticon anticon-github"
        role="img"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="github"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0138.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z"
          />
        </svg>
      </span>
       
      <a
        href="/repos/freeCodeCamp/freeCodeCamp"
        onClick={[Function]}
      >
        freeCodeCamp
        /
        freeCodeCamp
      </a>
    </h3>
    <p>
      freeCodeCamp.org's open source codebase and curriculum. Learn to code at home.
    </p>
    <div
      className="sc-AxirZ qsqry"
    >
      <p>
        <span
          aria-label="code"
          className="anticon anticon-code"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="code"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M516 673c0 4.4 3.4 8 7.5 8h185c4.1 0 7.5-3.6 7.5-8v-48c0-4.4-3.4-8-7.5-8h-185c-4.1 0-7.5 3.6-7.5 8v48zm-194.9 6.1l192-161c3.8-3.2 3.8-9.1 0-12.3l-192-160.9A7.95 7.95 0 00308 351v62.7c0 2.4 1 4.6 2.9 6.1L420.7 512l-109.8 92.2a8.1 8.1 0 00-2.9 6.1V673c0 6.8 7.9 10.5 13.1 6.1zM880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
            />
          </svg>
        </span>
         
        JavaScript
      </p>
      <p>
        <span
          aria-label="star"
          className="anticon anticon-star"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="star"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
            />
          </svg>
        </span>
         
        314,981
      </p>
      <p>
        <span
          aria-label="fork"
          className="anticon anticon-fork"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="fork"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M752 100c-61.8 0-112 50.2-112 112 0 47.7 29.9 88.5 72 104.6v27.6L512 601.4 312 344.2v-27.6c42.1-16.1 72-56.9 72-104.6 0-61.8-50.2-112-112-112s-112 50.2-112 112c0 50.6 33.8 93.5 80 107.3v34.4c0 9.7 3.3 19.3 9.3 27L476 672.3v33.6c-44.2 15-76 56.9-76 106.1 0 61.8 50.2 112 112 112s112-50.2 112-112c0-49.2-31.8-91-76-106.1v-33.6l226.7-291.6c6-7.7 9.3-17.3 9.3-27v-34.4c46.2-13.8 80-56.7 80-107.3 0-61.8-50.2-112-112-112zM224 212a48.01 48.01 0 0196 0 48.01 48.01 0 01-96 0zm336 600a48.01 48.01 0 01-96 0 48.01 48.01 0 0196 0zm192-552a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
            />
          </svg>
        </span>
         
        24,727
      </p>
    </div>
    <div
      className="sc-AxirZ qsqry"
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <a
        className="ant-btn ant-btn-primary"
        href="/repos/freeCodeCamp/freeCodeCamp"
        onClick={[Function]}
      >
        <span
          aria-label="right-circle"
          className="anticon anticon-right-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="right-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M666.7 505.5l-246-178A8 8 0 00408 334v46.9c0 10.2 4.9 19.9 13.2 25.9L566.6 512 421.2 617.2c-8.3 6-13.2 15.6-13.2 25.9V690c0 6.5 7.4 10.3 12.7 6.5l246-178c4.4-3.2 4.4-9.8 0-13z"
            />
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
          </svg>
        </span>
        <span>
          Details
        </span>
      </a>
    </div>
  </div>
</div>
`;
